libs/MeshKernel/src/CasulliRefinement.cpp

```diff
@@ -0,0 +1,724 @@
+        {
+            edgeId2 = edgeId1;
+        }
+    }
+    else
+    {
+        if (edgeId2 != constants::missing::uintValue)
+        {
+            edgeId1 = edgeId2;
+        }
+        else
+        {
+            throw AlgorithmError("Node edges specified: {}, {}", edgeId1, edgeId2);
+        }
+    }
+
+    UInt elementId = FindCommon(mesh, edgeId1, edgeId2);
+
+    if (elementId == constants::missing::uintValue)
+    {
+        throw AlgorithmError("No element found that shares edge: {} and {}", edgeId1, edgeId2);
+    }
+
+    UInt lr1 = IsLeftRight(mesh, elementId, edgeId1);
+    UInt lr2 = IsLeftRight(mesh, elementId, edgeId2);
+
+    UInt se1 = IsStartEnd(mesh, nodeId, edgeId1);
+    UInt se2 = IsStartEnd(mesh, nodeId, edgeId2);
+
+    if (edgeId1 == edgeId2)
+    {
+        lr1 = 1 - lr1;
+        lr2 = 1 - lr2;
+    }
+
+    UInt iPoint1 = se1 + 2 * (1 - lr1);
+    UInt iPoint2 = se2 + 2 * (1 - lr2);
+
+    if (newNodes[edgeId1][iPoint1] == constants::missing::uintValue)
+    {
+        newNodes[edgeId1][iPoint1] = newPointIndex;
+    }
+
+    if (newNodes[edgeId2][iPoint2] == constants::missing::uintValue)
+    {
+        newNodes[edgeId2][iPoint2] = newPointIndex;
+    }
+}
+
+meshkernel::UInt meshkernel::CasulliRefinement::IsStartEnd(const Mesh2D& mesh, const UInt nodeId, const UInt edgeId)
+{
+    UInt isStartEnd = constants::missing::uintValue;
+
+    if (mesh.m_edges[edgeId].first == nodeId)
+    {
+        isStartEnd = 0;
+    }
+    else if (mesh.m_edges[edgeId].second == nodeId)
+    {
+        isStartEnd = 1;
+    }
+
+    return isStartEnd;
+}
+
+meshkernel::UInt meshkernel::CasulliRefinement::IsLeftRight(const Mesh2D& mesh, const UInt elementId, const UInt edgeId)
+{
+    UInt isLeftRight = constants::missing::uintValue;
+    UInt edgeIndex = constants::missing::uintValue;
+    UInt nextEdgeIndex = constants::missing::uintValue;
+    UInt endNodeIndex = mesh.m_edges[edgeId].second;
+
+    for (UInt i = 0; i < mesh.m_facesEdges[elementId].size(); ++i)
+    {
+        UInt faceEdgeId = mesh.m_facesEdges[elementId][i];
+
+        if (faceEdgeId == edgeId)
+        {
+            edgeIndex = i;
+        }
+        else if (mesh.m_edges[faceEdgeId].first == endNodeIndex || mesh.m_edges[faceEdgeId].second == endNodeIndex)
+        {
+            nextEdgeIndex = i;
+        }
+    }
+
+    if (edgeIndex == constants::missing::uintValue || nextEdgeIndex == constants::missing::uintValue)
+    {
+        // EdgeId was not found
+        return isLeftRight;
+    }
+
+    if (nextEdgeIndex == edgeIndex + 1 || nextEdgeIndex + mesh.m_numFacesNodes[elementId] == edgeIndex + 1)
+    {
+        isLeftRight = 0;
+    }
+    else if (edgeIndex == nextEdgeIndex + 1 || edgeIndex + mesh.m_numFacesNodes[elementId] == nextEdgeIndex + 1)
+    {
+        isLeftRight = 1;
+    }
+
+    return isLeftRight;
+}
+
+meshkernel::UInt meshkernel::CasulliRefinement::FindCommon(const Mesh2D& mesh, const UInt edge1, const UInt edge2)
```

Comment:
Should this be moved to Mesh2D?

Reply:
Its been moved already, together with a couple of other functions.